File: app/schemas/visuales.py
"""Pydantic schemas for visualization endpoints (FASE 2)."""

from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class CapituloRatioResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    capitulo: str
    descripcion: Optional[str]
    minimo: Optional[float]
    percentil_25: Optional[float] = None
    mediana: Optional[float]
    percentil_75: Optional[float] = None
    maximo: Optional[float]
    desviacion_std: Optional[float] = None
    cantidad_datos: int
    estado_confiabilidad: Literal["muy_solido", "solido", "debil", "muy_debil"]
    building_type: Optional[str] = None


class ItemRatioResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    item_master_id: int
    item_key: str
    categoria: str
    muestras_total: int
    min_unitario: float
    p25_unitario: float
    median_unitario: float
    p75_unitario: float
    max_unitario: float
    avg_unitario: float


class ItemPresupuesto(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    capitulo: str = Field(..., description="Codigo de capitulo")
    valor_unitario: float = Field(..., gt=0, description="Precio unitario (EUR/m2)")
    cantidad: int = Field(default=1, ge=1, description="Cantidad")
    unidad: str = Field(default="m2", description="Unidad")

    @field_validator("capitulo", "unidad")
    @classmethod
    def validate_non_empty_string(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("no puede estar vacio")
        return normalized


class PresupuestoAnalisis(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    items: List[ItemPresupuesto] = Field(..., min_length=1)
    area_total: float = Field(..., gt=0, description="Area total en m2")
    building_type: Optional[str] = None

    @field_validator("building_type")
    @classmethod
    def normalize_building_type(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class ComparativaCapitulo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    capitulo: str
    descripcion: Optional[str]
    valor_mio: float = Field(..., description="Valor usuario (EUR/m2)")
    valor_ratio: float = Field(..., description="Mediana historica (EUR/m2)")
    desviacion_pct: float = Field(..., description="Desviacion en %")
    impacto_monetario: float = Field(..., description="Impacto total en EUR")
    estado_confiabilidad: Literal["muy_solido", "solido", "debil", "muy_debil"]
    ratio_encontrado: bool = True


class ResumenComparativa(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total_presupuesto: float
    total_ratio: float
    diferencia_pct: float
    diferencia_monetaria: float
    area_total: float
    confiabilidad_global: Literal["muy_solido", "solido", "debil", "muy_debil"]


class ComparativaResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    capitulos: List[ComparativaCapitulo]
    capitulos_sin_ratio: List[str] = Field(default_factory=list)
    resumen: ResumenComparativa

File: frontend/src/types/visuales.ts
export type EstadoConfiabilidad =
  | 'muy_solido'
  | 'solido'
  | 'debil'
  | 'muy_debil';

export interface CapituloRatioResponse {
  capitulo: string;
  descripcion?: string | null;
  minimo?: number | null;
  percentil_25?: number | null;
  mediana?: number | null;
  percentil_75?: number | null;
  maximo?: number | null;
  desviacion_std?: number | null;
  cantidad_datos: number;
  estado_confiabilidad: EstadoConfiabilidad;
  building_type?: string | null;
}

export interface ItemPresupuesto {
  capitulo: string;
  valor_unitario: number;
  cantidad: number;
  unidad: string;
}

export interface PresupuestoAnalisis {
  items: ItemPresupuesto[];
  area_total: number;
  building_type?: string | null;
}

export interface ComparativaCapitulo {
  capitulo: string;
  descripcion?: string | null;
  valor_mio: number;
  valor_ratio: number;
  desviacion_pct: number;
  impacto_monetario: number;
  estado_confiabilidad: EstadoConfiabilidad;
  ratio_encontrado: boolean;
}

export interface ResumenComparativa {
  total_presupuesto: number;
  total_ratio: number;
  diferencia_pct: number;
  diferencia_monetaria: number;
  area_total: number;
  confiabilidad_global: EstadoConfiabilidad;
}

export interface ComparativaResponse {
  capitulos: ComparativaCapitulo[];
  capitulos_sin_ratio: string[];
  resumen: ResumenComparativa;
}

export interface RangoResponse {
  chapter: string;
  items_count: number;
  muestras_total: number;
  min_unitario: number;
  max_unitario: number;
  p25_unitario: number;
  median_unitario: number;
  p75_unitario: number;
  avg_unitario: number;
}

export interface ItemRatioResponse {
  item_master_id: number;
  item_key: string;
  categoria: string;
  muestras_total: number;
  min_unitario: number;
  p25_unitario: number;
  median_unitario: number;
  p75_unitario: number;
  max_unitario: number;
  avg_unitario: number;
}

export interface Item {
  id: number;
  item_key: string;
  descripcion: string;
  categoria?: string | null;
  categoria_asignada: string;
  gama_asignada: string;
  muestras_count: number;
  ratio_actual?: number;
}
